### PR TITLE
feat: containerise screenshot.py

### PR DESCRIPTION
**Pull request by jamesericdavidson**
Run `screenshot.py` from a declarative `Containerfile`.

After this PR, we can work on the GitHub action: #184.

- [x] Create a `Containerfile` (FROM Ubuntu, Alpine, etc.; which can be kept as a runner, or used as a one-shot and disposed of)
  - [x] Install `firefox`, `python3`, and other dependencies (via the package manager)
- [x] Test `screenshot.py` from the container
  - [x] Fix system-specific breakages (e.g. macOS Firefox path)
  - [x] Move resultant files out of container/into working dir (`cp`)
  - [x] Fix monospace font support
- [x] ~~Installation and~~ Usage instructions

### CODE REVIEW

**Comment by jamesericdavidson** [comment]
Working via Podman on my machine. Latest run on `6b0755c89294cca9ee1e34b7c186628f40656a46`:

```sh
podman build .containers/
STEP 1/7: FROM python:alpine
STEP 2/7: WORKDIR /usr/src/app
--> Using cache 75bda4f7a499ea56dfa3f39edee5c740f05267a4056a3604471079d6cd03db00
--> 75bda4f7a499
STEP 3/7: COPY requirements.txt ./
--> Using cache 6b43df18abd6f28fd4804515db52f79a817187d04cca6b87e967b23f2d235410
--> 6b43df18abd6
STEP 4/7: RUN pip install --no-cache-dir -r requirements.txt
--> Using cache 0e2a0b075c79355300b31c9d1109e82e538f38fc0257c2582c47fd3c20607956
--> 0e2a0b075c79
STEP 5/7: RUN apk add firefox geckodriver
fetch https://dl-cdn.alpinelinux.org/alpine/v3.20/main/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.20/community/x86_64/APKINDEX.tar.gz
(1/121) Installing libSvtAv1Enc (2.0.0-r1)
(2/121) Installing aom-libs (3.9.1-r0)
(3/121) Installing libxau (1.0.11-r4)
(4/121) Installing libmd (1.1.0-r0)
(5/121) Installing libbsd (0.12.2-r0)
(6/121) Installing libxdmcp (1.1.5-r1)
(7/121) Installing libxcb (1.16.1-r0)
(8/121) Installing libx11 (1.8.9-r1)
(9/121) Installing hwdata-pci (0.382-r0)
(10/121) Installing libpciaccess (0.18.1-r0)
(11/121) Installing libdrm (2.4.120-r0)
(12/121) Installing libxext (1.3.6-r2)
(13/121) Installing libxfixes (6.0.1-r4)
(14/121) Installing wayland-libs-client (1.22.0-r4)
(15/121) Installing libva (2.21.0-r0)
(16/121) Installing libvdpau (1.5-r3)
(17/121) Installing libgcc (13.2.1_git20240309-r0)
(18/121) Installing libstdc++ (13.2.1_git20240309-r0)
(19/121) Installing onevpl-libs (2023.3.1-r2)
(20/121) Installing ffmpeg-libavutil (6.1.1-r8)
(21/121) Installing libdav1d (1.4.2-r0)
(22/121) Installing openexr-libiex (3.1.13-r1)
(23/121) Installing openexr-libilmthread (3.1.13-r1)
(24/121) Installing imath (3.1.12-r0)
(25/121) Installing openexr-libopenexr (3.1.13-r1)
(26/121) Installing brotli-libs (1.1.0-r2)
(27/121) Installing giflib (5.2.2-r0)
(28/121) Installing libhwy (1.0.7-r0)
(29/121) Installing libjpeg-turbo (3.0.3-r0)
(30/121) Installing lcms2 (2.16-r0)
(31/121) Installing libpng (1.6.44-r0)
(32/121) Installing libjxl (0.10.2-r0)
(33/121) Installing lame-libs (3.100-r5)
(34/121) Installing opus (1.5.2-r0)
(35/121) Installing rav1e-libs (0.7.1-r0)
(36/121) Installing libgomp (13.2.1_git20240309-r0)
(37/121) Installing soxr (0.1.3-r7)
(38/121) Installing ffmpeg-libswresample (6.1.1-r8)
(39/121) Installing libogg (1.3.5-r5)
(40/121) Installing libtheora (1.1.1-r18)
(41/121) Installing libvorbis (1.3.7-r2)
(42/121) Installing libvpx (1.14.1-r0)
(43/121) Installing libsharpyuv (1.3.2-r0)
(44/121) Installing libwebp (1.3.2-r0)
(45/121) Installing libwebpmux (1.3.2-r0)
(46/121) Installing x264-libs (0.164_git20231001-r0)
(47/121) Installing numactl (2.0.18-r0)
(48/121) Installing x265-libs (3.6-r0)
(49/121) Installing xvidcore (1.3.7-r2)
(50/121) Installing ffmpeg-libavcodec (6.1.1-r8)
(51/121) Installing mesa (24.0.9-r1)
(52/121) Installing libexpat (2.6.3-r0)
(53/121) Installing wayland-libs-server (1.22.0-r4)
(54/121) Installing mesa-gbm (24.0.9-r1)
(55/121) Installing mesa-glapi (24.0.9-r1)
(56/121) Installing libxshmfence (1.3.2-r6)
(57/121) Installing mesa-egl (24.0.9-r1)
(58/121) Installing libxxf86vm (1.1.5-r6)
(59/121) Installing mesa-gl (24.0.9-r1)
(60/121) Installing pciutils-libs (3.12.0-r1)
(61/121) Installing libxcomposite (0.4.6-r5)
(62/121) Installing libxdamage (1.1.6-r5)
(63/121) Installing libxrender (0.9.11-r5)
(64/121) Installing libxrandr (1.5.4-r1)
(65/121) Installing alsa-lib (1.2.11-r0)
(66/121) Installing libintl (0.22.5-r0)
(67/121) Installing libeconf (0.6.3-r0)
(68/121) Installing libblkid (2.40.1-r1)
(69/121) Installing libmount (2.40.1-r1)
(70/121) Installing pcre2 (10.43-r0)
(71/121) Installing glib (2.80.5-r0)
(72/121) Installing libatk-1.0 (2.52.0-r0)
(73/121) Installing freetype (2.13.2-r0)
(74/121) Installing fontconfig (2.15.0-r1)
(75/121) Installing pixman (0.43.2-r0)
(76/121) Installing cairo (1.18.0-r0)
(77/121) Installing cairo-gobject (1.18.0-r0)
(78/121) Installing dbus-libs (1.14.10-r1)
(79/121) Installing libevent (2.1.12-r7)
(80/121) Installing pkgconf (2.2.0-r0)
(81/121) Installing libxml2 (2.12.7-r0)
(82/121) Installing shared-mime-info (2.4-r0)
(83/121) Installing hicolor-icon-theme (0.17-r2)
(84/121) Installing zstd-libs (1.5.6-r0)
(85/121) Installing tiff (4.6.0t-r0)
(86/121) Installing gdk-pixbuf (2.42.12-r0)
(87/121) Installing gtk-update-icon-cache (3.24.43-r0)
(88/121) Installing libxcursor (1.2.2-r1)
(89/121) Installing libxi (1.8.1-r4)
(90/121) Installing libxinerama (1.1.5-r4)
(91/121) Installing libxtst (1.2.4-r6)
(92/121) Installing at-spi2-core (2.52.0-r0)
(93/121) Installing libatk-bridge-2.0 (2.52.0-r0)
(94/121) Installing avahi-libs (0.8-r17)
(95/121) Installing gmp (6.3.0-r1)
(96/121) Installing nettle (3.9.1-r0)
(97/121) Installing libunistring (1.2-r0)
(98/121) Installing libidn2 (2.3.7-r0)
(99/121) Installing libtasn1 (4.19.0-r2)
(100/121) Installing p11-kit (0.25.3-r0)
(101/121) Installing gnutls (3.8.5-r0)
(102/121) Installing cups-libs (2.4.9-r1)
(103/121) Installing libepoxy (1.5.10-r1)
(104/121) Installing fribidi (1.0.15-r0)
(105/121) Installing graphite2 (1.3.14-r6)
(106/121) Installing harfbuzz (8.5.0-r0)
(107/121) Installing libxft (2.3.8-r3)
(108/121) Installing pango (1.52.2-r0)
(109/121) Installing wayland-libs-cursor (1.22.0-r4)
(110/121) Installing wayland-libs-egl (1.22.0-r4)
(111/121) Installing xkeyboard-config (2.41-r0)
(112/121) Installing libxkbcommon (1.7.0-r0)
(113/121) Installing gtk+3.0 (3.24.43-r0)
(114/121) Installing icu-data-en (74.2-r0)
Executing icu-data-en-74.2-r0.post-install
*
* If you need ICU with non-English locales and legacy charset support, install
* package icu-data-full.
*
(115/121) Installing icu-libs (74.2-r0)
(116/121) Installing nspr (4.35-r4)
(117/121) Installing nss (3.104-r0)
(118/121) Installing scudo-malloc (17.0.6-r0)
(119/121) Installing libwebpdemux (1.3.2-r0)
(120/121) Installing firefox (131.0.2-r0)
(121/121) Installing geckodriver (0.34.0-r0)
Executing busybox-1.36.1-r29.trigger
Executing glib-2.80.5-r0.trigger
Executing shared-mime-info-2.4-r0.trigger
Executing gdk-pixbuf-2.42.12-r0.trigger
Executing gtk-update-icon-cache-3.24.43-r0.trigger
Executing gtk+3.0-3.24.43-r0.trigger
OK: 367 MiB in 148 packages
--> 2ffa900e569c
STEP 6/7: COPY . .
--> 1031085c43d9
STEP 7/7: CMD [ "python", "./screenshot.py" ]
COMMIT
--> bd72e76bd3eb
bd72e76bd3eb0c19e2aff61777db4692f0ac5dab45620019b960d1622810f8f0

gokarna on  feat/screenshot-container [$] via 🐍 v3.13.0 (screenshot) 
❯ podman run bd72e76bd3eb
Traceback (most recent call last):
  File "/usr/src/app/./screenshot.py", line 104, in <module>
    main()
    ~~~~^^
  File "/usr/src/app/./screenshot.py", line 97, in main
    take_screenshots(driver, base_url, SCREENSHOT_OPTIONS)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/app/./screenshot.py", line 37, in take_screenshots
    driver.get(base_url + opt['path'])
    ~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/site-packages/selenium/webdriver/remote/webdriver.py", line 355, in get
    self.execute(Command.GET, {"url": url})
    ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/site-packages/selenium/webdriver/remote/webdriver.py", line 346, in execute
    self.error_handler.check_response(response)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^
  File "/usr/local/lib/python3.13/site-packages/selenium/webdriver/remote/errorhandler.py", line 245, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.WebDriverException: Message: Reached error page: about:neterror?e=connectionFailure&u=http%3A//127.0.0.1%3A1313/&c=UTF-8&d=Firefox%20can%E2%80%99t%20establish%20a%20connection%20to%20the%20server%20at%20127.0.0.1%3A1313.
Stacktrace:
RemoteError@chrome://remote/content/shared/RemoteError.sys.mjs:8:8
WebDriverError@chrome://remote/content/shared/webdriver/Errors.sys.mjs:193:5
UnknownError@chrome://remote/content/shared/webdriver/Errors.sys.mjs:832:5
checkReadyState@chrome://remote/content/marionette/navigate.sys.mjs:58:24
onNavigation@chrome://remote/content/marionette/navigate.sys.mjs:343:39
emit@resource://gre/modules/EventEmitter.sys.mjs:148:20
receiveMessage@chrome://remote/content/marionette/actors/MarionetteEventsParent.sys.mjs:33:25
```

**Comment by jamesericdavidson** [comment]
Reproducible in `distrobox` for local testing (i.e. avoid rebuilds):

```sh
cd path/to/gokarna
distrobox create -i python:alpine
distrobox enter python-alpine
pip install --no-cache-dir -r .containers/requirements.txt
sudo apk add geckodriver firefox
python .containers/screenshot.py
```

File: Containerfile [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# https://hub.docker.com/_/python/
+
```
Comment (yashmehrotra) [inline Containerfile:2]: (nitpick): No need for this comment

**Comment by jamesericdavidson** [comment]
Font/emoji support for Alpine:

https://wiki.alpinelinux.org/wiki/Fonts
https://wiki.alpinelinux.org/wiki/Emojis

**Comment by jamesericdavidson** [comment]
@526avijitgupta @yashmehrotra Fixed `screenshot.py` and added a little build script. Please test locally and see how it works for you - this is an interim solution before looking at GitHub workflows/Actions.

P.S. I looked into [updating existing images](https://stackoverflow.com/questions/32112035/docker-cp-doesnt-copy-file-into-container/32126677#32126677), but found the solution messy. GitHub Actions wouldn't be stateful anyway, so I'm not considering adding it back in.

**Comment by jamesericdavidson** [comment]
Last minute notes:

- Docker and Podman work on my machine; the script defaults to using `docker`
  - You need a `docker` or `podman` binary in your `$PATH`
- The `dash` shebang is used for speed, and POSIX-compliance
  `dash` is pre-installed on Mac, Debian and derivatives (Ubuntu, Linux Mint, etc.)

**Comment by yashmehrotra** [comment]
@jamesericdavidson I think I'll merge this for now, have some ideas on how to make it part of the CI flow which I'll probably implement over the weekend

**Pull request merged**